Integrate UCAM database API

davidverweij opened the pull request:
Closes #45 

_Ps. I had to squash my commits as rebasing gave some headaches_

This branch introduces the integration of the UCAM api based on a test API endpoint provided to us. Changes include:
- a new `consumer` endpoint called `ucam` including 6 sub-endpoints
- a different schema for `ucam` 'records'. Before, these were based on rows in a .csv - now these are based on Patients or Devices, with a sub-list or linked Devices or Patients (Patient-_Devices_ and Device-_Patients_) respectively. 
- the `ucam` endpoints implements authentication similar to the [dmpy](https://github.com/ideafast/dmpy/blob/master/dmpy/client.py) by reading and writing into the environment. This allows retrieval of auth across pipelines.
- an added test-suite to test the `consumer` implementation of this endpoint as well as the `data_transfer` implementation. The first focuses on the API, the latter mainly focuses on the additional methods in `data_transfer` that have some logic to determine patients or IDs based on wear times.
  - these test includes integration test, marked with `pytest.mark.live` - and are excluded from any standard `nox` session. Instead, you can run these manually: `pytest -vm live`

## Testing
```shell
git pull
git checkout ucam_api_2
poetry install
poetry shell
poetry run consumer
```
In your browser, navigate to
- [http://0.0.0.0:8000/ucam/devices](http://0.0.0.0:8000/ucam/devices)
- [http://0.0.0.0:8000/ucam/patients](http://0.0.0.0:8000/ucam/patients)
- [http://0.0.0.0:8000/ucam/vtt](http://0.0.0.0:8000/ucam/vtt)
- [http://0.0.0.0:8000/ucam/patients/WRONG_ID](http://0.0.0.0:8000/ucam/patients/WRONG_ID)
- etc..

```shell
# in another terminal instance
poetry run nox -r
pytest -vm live

# remove the upload step from shared jobs
python data_transfer/main.py BTF Kiel 0 1  # increase last number if no results
```

davidverweij commented:
Added a statement to ignore any UCAM records if they contain deviations. This has to be temporarily used, as currently (once live) UCAM data contains wrongly associated device IDs; since the correct IDs were not available for selection. Running the pipeline where these mis-links are present will cause wrongfully linked data to wrong patients. 

## Test
Run `pytest -vsk "test_get_patient_by_period_date_ignore_deviations"`
or `poetry run nox -rs tests`

jawrainey commented on consumer/services/ucam.py:
```diff
@@ -0,0 +1,121 @@
+    return [as_devicepatient(vtt) for vtt in payload] if payload else None
+
+
+def as_patient(p: dict) -> Patient:
```
re:routers and models on consumer. Rather than us having a `serialize` method we could leverge fast_api to do that for us:

- https://fastapi.tiangolo.com/tutorial/response-model/
- [other example](https://github.com/nsidnev/fastapi-realworld-example-app/blob/f3830df35643b39b723991af4310f552b8e0b668/app/api/routes/comments.py#L28) 

davidverweij commented on consumer/services/ucam.py:
```diff
@@ -0,0 +1,121 @@
+    response = requests.get(url, headers=headers)
+    response.raise_for_status()
+
+    # possibly no result
+    if response.status_code == 204:
+        return None
```
I have requested, but no response yet. The difference is most likely since patients are unique and (unfortunately due to the past use of UCAM) devices and VTT are not. But either way - have requested it but we can work with this at the moment.

davidverweij commented on data_transfer/config.py:
```diff
@@ -38,12 +39,11 @@ class GlobalConfig(BaseSettings):
     database_name: str = "pipeline_local"
 
     inventory_api: str = ""
+    ucam_api: str = ""
```
How come this should be in the `Settings`, but `inventory_api` does not?

jawrainey replied:
@davidverweij -- you're right. Probs best to move `inventory_api` to `Settings` too 🚀

jawrainey replied:
Worth noting that the mocking of the inventory tests are not ideal because they depend on method internals e.g., `test_record_by_device_id_history_is_none` mocks `requests.get` when it should mock `device_history` (where requests.get is used) and remove the dependency on inventory_api, e.g., 

from: 
```python
    response.json = lambda: mock_inventory_history_response

    with patch("requests.get", return_value=response):
```
to
```python
    response.json = lambda: mock_inventory_history_response['data']

    with patch.object(inventory, "device_history", return_value=response):
```

davidverweij replied:
So, when I move inventory (and thus ucam as well) from the `GlobalConfig` to `Settings`, test won't run (due to the `env` TESTING="True") and the `inventory_api` and `ucam_api` is not set. Thoughts?

davidverweij commented on data_transfer/services/ucam.py:
```diff
@@ -1,135 +1,140 @@
-    A device may be worn by many patients. This returns such history.
-    """
-    return [r for r in __get_patients() if r.device_id == device_id]
+                # NOTE: ignores any device ID if record contains deviations
+                # TODO: removes this constraint once faulty device_id assignment has been resolved
+                if patient.deviations:
+                    continue
```
Done: https://github.com/ideafast/middleware-services/issues/81

davidverweij commented on tests/data_transfer_tests/services/test_ucam.py:
```diff
@@ -1,46 +1,221 @@
+        assert all(isinstance(x, Device) for x in result)
+        assert len(result) == 1
+        assert result[0].device_id == "NR3-DEVICE"
+        assert all(isinstance(x, DevicePatient) for x in result[0].patients)
```
Have removed redundant checks, but haven't written tests to validate the schema mainly due to priorities at the moment.

davidverweij commented on consumer/config.py:
```diff
@@ -13,6 +13,11 @@ class Settings(BaseSettings):
     support_base_url: str = ""
     support_token: str = ""
 
+    # UCAM API
+    ucam_uri: str = ""
+    ucam_username: str = ""
+    ucam_password: str = ""
+
```
Have left them for now due to project priorities, but have extended issue #82

davidverweij commented on tests/consumer_tests/routers/test_ucam.py:
```diff
@@ -0,0 +1,148 @@
+    result = ucam.get_patients()
+
+    assert all(isinstance(x, Patient) for x in result)
+
```
I've worked on this a bit, but for some reason could not get pytest to pick up on the 'functional' folder. Also - as we discussed quite a while a go - I can't name the test subfolder the same name as the module folder, as then any module imports will fail.

In addition, I was refactoring a very large number of things here, and since this branch is not up to date with master (and overall this is getting really big now), I'd like to move this into a different issue and PR.

davidverweij commented:
@jawrainey, I've implemented _almost_ all of your feedback, but have had to push a few improvements into issues in order to keep this PR manageable and allow myself to work on other aspects of the pipeline (and our deadlines as is; in particular the refactoring of the tests and additional test structure. 

I'm happy to improve and work on those in the future.

jawrainey commented:
Tested locally with above changes and works perfectly 👍🏼 Thanks for opening the other issues and I agree that seperating these out into smaller, more managable PRs is ideal. Looking forward to going live once we have API access 🚀

jawrainey reviewed (approved):
Tested locally with above changes and works perfectly 👍🏼 Thanks for opening the other issues and I agree that seperating these out into smaller, more managable PRs is ideal. Looking forward to going live once we have API access 🚀

davidverweij commented:
@jawrainey, UCAM has now added the ByteFlies dots to the payload - which are joined for each device and _only_ in the `GET /patients/` endpoint. See [the mock data](https://github.com/ideafast/middleware-services/blob/ucam_api_2/tests/consumer_tests/routers/data/mock_ucam.json) for how that looks like. 

I  added a (temporary) endpoint [`GET /btf/`](https://github.com/ideafast/middleware-services/blob/5b2661a8170d39225164b76c427bb7da498301b3/consumer/services/ucam.py#L96-L139) that (under the hood) queries the `GET /patients/` endpoint, and iteratively converts any present BTF dots into a list of devices - to be returned as a payload identical to what we expect from `GET /devices`. It is then serialised as normal.

In the data_transfer, I added two methods to query this endpoint (which caches the result), and another one to do a lookup in that result (to get a single dot). 

I've tested locally with `python data_transfer/main.py BTF Newcastle 6 6` which resulted in `67 records created and 53 NOT this session`. Given these were recent I have high hopes for more historical data! I could check that a record that was being downloaded was not based on inventory and not found in the ucam `/devices` endpoint, but was found thanks to the `/btf` custom endpoint.

## Test
- pull down changes
- run `poetry run nox -rs tests`

### Test btf_dot code with live API
- assuming you have credentials for UCAM set up in your env
- run `poetry run consumer`
- In _tests/data_transfer_tests/services/test_ucam.py_ add the code below, and
- replace the `key` and `expected` values with existing ID's (see MSTeams for a suggestion)
- run `pytest -vsk "test_LIVE_get_btfdots"`
```python
import pytest

@pytest.mark.live
@pytest.mark.parametrize(
    "key, expected",
    [("XXX-XXXXXX", "XXX-XXXXXX"), ("YYY-YYYYYY", "YYY-YYYYYY")]
)
def test_LIVE_get_btfdots(key: str, expected: str) -> None:
    result = ucam.get_one_btf_dot(key)
    assert result[0].device_id == expected
    print(result)
```

## Thoughts?
Let me know what you think and then I feel confident to merge the changes and run the pipeline against - say - Muenster first.

_Note that any deviation is still ignored as I spotted some anomalies. I can ask all clinicians to report back to us when they finished QA so that we may run the pipeline again._

jawrainey commented:
Great work @davidverweij 🚀 As you have verified this PR through deploying a dev pipeline for BTF, which I also checked on the DMP and can confirm records are as expected. I have also verified the PR through testing locally and deploying a dev pipeline for DRM, which required merging master given my recent change due to a conflict. This also required building new images to test the deployment as master had a few changes.
